### PR TITLE
Add Out of Training Set Pipeline Evaluation (OOTSP)

### LINKED ISSUE #30: Predict on out-of-training-set pipelines
Some models cannot make predictions about pipelines which are not in the training set. (All pipelines must use primitives from the training set of course.) We need to show-case the models which can make predictions on novel pipelines.

To do this, the pipelines in the training data need to be restricted to some smaller subset. We should be careful that we do not eliminate any primitives from the training set. We would then have the models make predictions they can on the the full test set (with out-of-training-set pipelines). Those that can recommend pipelines from the full test set will likely have better scores, and vice-versa.

### PR DESCRIPTION
**Pull request by orionw**
- Support for testing a test set containing only out of training set pipelines
- Support for testing a the other part of that test set for validation purposes
- Configurable parameter for percentage split of OOTSP

Closes #30 

Note: PMF, although not refactored, still works on the OOTSP problem.  It will be refactored to so that it is more in line by not using the validation set to fit the one-hot encodings, but it works on both in-training-set-pipelines test set and the OOTSP test set

Also note: 
We should revert back main.sh after you review this (except for the `metafeature_subset` line I added for the full dataset)

### CODE REVIEW

File: main.sh [hunk 1]
```diff
@@ -26,20 +28,21 @@ results_dir=./dev_results
 python3 -m dna split-data \
     --data-path $raw_data_path \
     --test-size $test_size \
-    --split-seed $test_split_seed
+    --split-seed $test_split_seed
```
Comment (bjschoenfeld) [inline main.sh]: try not to add trailing whitespace

**Comment by bjschoenfeld** [comment]
I am getting errors in daglstm and hiddendaglstm.

**Comment by erikhuck** [comment]
> I am getting errors in daglstm and hiddendaglstm.

If I had to guess, this might be due to the preprocessing that the hiddendaglstm and daglstm models do to the data set. That preprocessing must also occur in the test set or whatever data set that those models use.

**Comment by orionw** [comment]
@erikhuck You are correct.  Because it preprocesses the data in `fit` only once, it never preprocesses the additional OOTSP test set and so it fails.  We want to move the preprocessing done in `fit` to the rnn dataloader function (since we pass the `self.pipeline_structure` into the dataloader already - let's just pass in the data and have it do the preprocessing there.  Then all datasets will be preprocessed.  I was going to do it, but if you have time feel free to add to this PR.

**Comment by erikhuck** [comment]
@orionw Brandon would like only one person to be working on a branch at a time. Are all your changes pushed? If so, I'll assign this PR to myself and take over from here.

**Comment by orionw** [comment]
@erikhuck You're good to go.  If you can do it today, go for it.  If you don't finish it, I can finish this afternoon.  I know you have some other PR's you're working on too, so whatever you want to work on is fine.   Worst case I can branch off of your branch and then merge it back in.

**Comment by erikhuck** [comment]
Are we certain that the out of training set pipelines don't contain primitives that are **not** found in the training set?

**Comment by orionw** [comment]
We are not certain.  I think Brandon is okay with that possibility of failure, but I’m not sure.  We should ask

**Comment by bjschoenfeld** [comment]
> Are we certain that the out of training set pipelines don't contain primitives that are not found in the training set?

Let's make this assumption. If it happens not to be true for some split, I want it to fail spectacularly. If you see in the code where an OOTS primitive would cause an error (e.g. key error because the primitive name is not in some dict), let's explicitly control it and raise an exception that indicates there was an OOTS primitive. You don't have to go out of your way to find these, just make the change if you are already working in that code block.

**Comment by erikhuck** [comment]
Hey Orion, I'm right in the middle of the process. Is it okay if I stay assigned to this and finish it tomorrow?

**Comment by erikhuck** [comment]
Scratch that. I'm going to work late until it's done.
